src/node.cc

```diff
@@ -3817,22 +3794,67 @@ int Start(int argc, char** argv) {
           more = true;
       }
     } while (more == true);
-    code = EmitExit(env);
+
+    int exit_code = EmitExit(env);
+    if (instance_data->is_main())
+      instance_data->set_exit_code(exit_code);
     RunAtExit(env);
 
     env->Dispose();
     env = nullptr;
   }
 
-  CHECK_NE(node_isolate, nullptr);
-  node_isolate->Dispose();
-  node_isolate = nullptr;
+  CHECK_NE(isolate, nullptr);
+  isolate->Dispose();
+  isolate = nullptr;
+
+  // Main node instance uses uv_default_loop.
+  if (instance_data->is_worker())
+    uv_loop_delete(instance_data->loop());
+}
+
+int Start(int argc, char** argv) {
+  PlatformInit();
+
+  const char* replaceInvalid = secure_getenv("NODE_INVALID_UTF8");
+
+  if (replaceInvalid == nullptr)
+    WRITE_UTF8_FLAGS |= String::REPLACE_INVALID_UTF8;
+
+  CHECK_GT(argc, 0);
+
+  // Hack around with the argv pointer. Used for process.title = "blah".
+  argv = uv_setup_args(argc, argv);
+
+  // This needs to run *before* V8::Initialize().  The const_cast is not
+  // optional, in case you're wondering.
+  int exec_argc;
+  const char** exec_argv;
+  Init(&argc, const_cast<const char**>(argv), &exec_argc, &exec_argv);
+
+#if HAVE_OPENSSL
+  // V8 on Windows doesn't have a good source of entropy. Seed it from
+  // OpenSSL's pool.
+  V8::SetEntropySource(crypto::EntropySource);
+#endif
+
+  V8::InitializePlatform(new Platform(4));
+  V8::Initialize();
+
+  NodeInstanceData instance_data(NodeInstanceType::MAIN,
+                                      uv_default_loop(),
+                                      argc,
+                                      const_cast<const char**>(argv),
+                                      exec_argc,
+                                      exec_argv);
+  StartNodeInstance(&instance_data);
+  int exit_code = instance_data.exit_code();
   V8::Dispose();
 
   delete[] exec_argv;
   exec_argv = nullptr;
```

Comment:
There is a potential use-after-free waiting to happen here in that exec_argv has been freed now but instance_data still retains a pointer.  I suggest putting instance_data and the call to StartNodeInstance() in a block scope to make that impossible.